src/Simulation/DocumentationGenerationStep/Extensions.cs

```diff
@@ -0,0 +1,246 @@
+    {
+        private QsCustomType type;
+        internal Udt(QsCustomType type)
+        {
+            this.type = type;
+        }
+
+        public QsNullable<QsLocation> Location => type.Location;
+
+        public IAttributeBuilder<QsCustomType> AddAttribute(QsDeclarationAttribute attribute) =>
+            new Udt(type.AddAttribute(attribute));
+
+        public QsCustomType Build() => type;
+    }
+
+    internal static class Extensions
+    {
+        internal static IAttributeBuilder<QsCallable> AttributeBuilder(
+            this QsCallable callable
+        ) => new Callable(callable);
+        internal static IAttributeBuilder<QsCustomType> AttributeBuilder(
+            this QsCustomType type
+        ) => new Udt(type);
+
+        internal static IAttributeBuilder<T> WithAttribute<T>(
+            this IAttributeBuilder<T> builder, string @namespace, string name,
+            TypedExpression input
+        ) =>
+            builder.AddAttribute(
+                new QsDeclarationAttribute(
+                    QsNullable<UserDefinedType>.NewValue(
+                        new UserDefinedType(
+                            NonNullable<string>.New(@namespace),
+                            NonNullable<string>.New(name),
+                            QsNullable<Tuple<QsPositionInfo, QsPositionInfo>>.Null
+                        )
+                    ),
+                    input,
+                    builder.Location.Item.Offset,
+                    QsComments.Empty
+                )
+            );
+
+        private static IAttributeBuilder<T> WithDocumentationAttribute<T>(
+            this IAttributeBuilder<T> builder, string attributeName,
+            TypedExpression input
+        ) => builder.WithAttribute("Microsoft.Quantum.Documentation", attributeName, input);
+
+        private static TypedExpression AsLiteralExpression(this string literal) =>
+            SyntaxGenerator.StringLiteral(
+                NonNullable<string>.New(literal),
+                ImmutableArray<TypedExpression>.Empty
+            );
+
+        internal static IAttributeBuilder<T> MaybeWithSimpleDocumentationAttribute<T>(
+            this IAttributeBuilder<T> builder, string attributeName, string? value
+        ) =>
+            value == null || value.Trim().Length == 0
+            ? builder
+            : builder.WithDocumentationAttribute(
+                attributeName, value.AsLiteralExpression()
+            );
+
+        internal static IAttributeBuilder<T> WithListOfDocumentationAttributes<T>(
+            this IAttributeBuilder<T> builder, string attributeName, IEnumerable<string> items
+        ) =>
+            items
+            .Aggregate(
+                builder,
+                (acc, item) => acc.WithDocumentationAttribute(
+                    attributeName, item.AsLiteralExpression()
+                )
+            );
+
+        internal static IAttributeBuilder<T> WithDocumentationAttributesFromDictionary<T>(
```

Comment:
I believe you might want to use the tools in this PR: https://github.com/microsoft/qsharp-compiler/pull/463
Let me know what attribute related things are missing, and we can see if it makes sense to add it to that static class.

Reply:
Sounds good, I'll take a look at that, then. One thing I ran into was that a lot of the documentation attributes take a similar form on UDTs and callables, but even though QsCallable and QsCustomType both have an `AddAttribute` method, there's no common interface that allows calling `AddAttribute` on both. The `AttributeBuilder` class here was a hack around that to reduce code duplication.